### PR TITLE
docs: add section on local cloud-init dataource

### PR DESCRIPTION
**Pull request by codyshepherd**
Refs:
- https://warthogs.atlassian.net/browse/CPC-4698

### CODE REVIEW

**Comment by j5awry** [comment]
holy sphinx upstream issues batman!

```
TypeError: unsupported operand type(s) for /: 'str' and 'str'
```

not sure if this is going to cause failures to build generally

**Comment by k-dimple** [comment]
Yeah, it is causing general build failures for our code. I am looking into it. (We can avoid the error if we decide to remove all images from our code! :) .. but we are not going to do that)

File: public-images/public-images-how-to/index.rst [hunk 1]
```diff
@@ -2,10 +2,11 @@
 
 How-to guides
 =============
-These guides provide instructions for performing basic operations related to Vagrant.
+These guides provide instructions for performing basic operations related to public Ubuntu cloud images.
 
 - :ref:`vagrant-bartender`
 - :ref:`run-a-vagrant-box`
+- :ref:`use-local-ds`
```
Comment (k-dimple) [inline public-images/public-images-how-to/index.rst:9]: At the end of this file (under the section toctree section) add: 
`Use a local cloud-init datasource <use-local-ds>`

File: public-images/public-images-how-to/use-local-cloud-init-ds.rst [hunk 2]
```diff
@@ -0,0 +1,96 @@
+.. _use-local-ds:
+
+Create and Use a Local Cloud-Init Datasource
```
Comment (k-dimple) [inline public-images/public-images-how-to/use-local-cloud-init-ds.rst:3]:
```suggestion
Create and use a local cloud-init datasource
```

File: public-images/public-images-how-to/use-local-cloud-init-ds.rst [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+============================================
+
+Purpose
+-------
+
+The steps in this How-To guide will allow you to create a seed image that defines a custom cloud-init datasource.
```
Comment (k-dimple) [inline public-images/public-images-how-to/use-local-cloud-init-ds.rst:9]:
```suggestion
The steps in this how-to guide will allow you to create an Ubuntu seed image that defines a custom cloud-init datasource.
```
Does the above suggestion make sense? When you say seed image, I usually assume something like a core Ubuntu image that can be used as the starting point for adding further packages or making changes that customise it.

File: public-images/public-images-how-to/use-local-cloud-init-ds.rst [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+Using a custom local datasource is a great way to inject custom user data or metadata into cloud images launched
+locally, e.g. with `qemu`.
+
```
Comment (k-dimple) [inline public-images/public-images-how-to/use-local-cloud-init-ds.rst:13]: Can we add one more line here about why one might want to do this, i.e. launch such customised images locally? Is it for some sort of testing?

File: public-images/public-images-how-to/use-local-cloud-init-ds.rst [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    #cloud-config
+    password: password
+    chpasswd:
+      expire: False
+    ssh_pwauth: True
+    ssh_authorized_keys:
+      - ssh-rsa AAAA...UlIsqdaO+w==
+    EOF
+
+Datasource metadata (optional)
```
Comment (k-dimple) [inline public-images/public-images-how-to/use-local-cloud-init-ds.rst:45]:
```suggestion
Specify datasource metadata (optional)
```

**Review by k-dimple — CHANGES_REQUESTED** [review]

File: public-images/public-images-how-to/use-local-cloud-init-ds.rst [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+-----------------------------
+
+If desired, create a YAML file with your desired datasource `metadata`_. Metadata describes typically cloud-defined
```
[the remark this answers is not in the document]
Comment (codyshepherd) [reply public-images/public-images-how-to/use-local-cloud-init-ds.rst:48]: I disagree with this suggestion. The metadata is typically cloud-defined (except in cases like this when there's not an actual cloud). Metadata _always_ describes this data.

File: public-images/public-images-how-to/use-local-cloud-init-ds.rst [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+Download the latest Ubuntu server image, then pass that and your newly-created seed image to the qemu launch command.
+
+The following launches a cloud image with:
+
+* KVM acceleration
+* the local machine's CPUs
+* 2GB of memory
+* no graphics and serial output to the console
+* snapshot will make writes to a temporary file instead of the disk image itself. This ensures the base disk is not touched. If at some point, you want to persist the changes you've made on the disk, press C-a s
+* virtio network device that redirects guest port 22 to host's port 2222
```
[the remark this answers is not in the document]
Comment (codyshepherd) [reply public-images/public-images-how-to/use-local-cloud-init-ds.rst:76]: That's correct. This is pretty standard terminology in the context of virtualization/containerization. I'll add a bit of extra text.

File: public-images/public-images-how-to/use-local-cloud-init-ds.rst [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+* virtio cloud image
+* virtio seed image
```
[the remark this answers is not in the document]
Comment (codyshepherd) [reply public-images/public-images-how-to/use-local-cloud-init-ds.rst:78]: It is a bit confusing, but it kinda feels beyond the scope of this article, and that's what the reference links to QEMU docs are for. The [seed image](https://cloudinit.readthedocs.io/en/latest/howto/run_cloud_init_locally.html#create-an-iso-disk) is just a disk image (iso) that contains the user data and metadata cloud-init needs to do its job. I'll add some text.

File: public-images/public-images-how-to/use-local-cloud-init-ds.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (k-dimple) [reply public-images/public-images-how-to/use-local-cloud-init-ds.rst:48]: Ah now I see what you mean! Should we just remove 'typically' from the statement? It does seem like a typo on first reading!

**Review by k-dimple — APPROVED** [review]

**Pull request merged**